### ISSUE TITLE
Uploading S3 files to buckets in multiple regions and of multiple types

### ISSUE THREAD

**Issue by troy** [aws/aws-sdk-ruby#510]
I encountered a case that I'd like understand and document in the readme, wiki, or a blog post. Here goes.

## Problem

I'm uploading files to user-supplied bucket names which can exist in any region. In the process, I encountered two challenges. I've worked through one and am stuck on the second. They're below.

## Handling PermanentRedirects, part 1

By default, `aws-sdk` requests are made to an endpoint for the `us-east-1` region. When one calls `S3Object#put`, S3 may respond with a `PermanentRedirect` to an endpoint in another region. 

This is a HTTP 301 which does **not** include a `Location` header (see [transcript](http://docs.aws.amazon.com/AmazonS3/latest/dev/Redirects.html#RedirectsPermanentRedirection) from a similar-looking temporary redirect). The response body looks like this, with newlines and indents added for readability:

```
<?xml version="1.0" encoding="UTF-8"?>
  <Error>
    <Code>PermanentRedirect</Code>
    <Message>The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.</Message>
    <RequestId>abcdefg</RequestId>
    <Bucket>bucket.example.com</Bucket>
    <HostId>1a2b3c</HostId>
    <Endpoint>bucket.example.com.s3.amazonaws.com</Endpoint>
  </Error>
```

As of this writing, in a default config, this will manifest as an `AWS::S3::Errors::PermanentRedirect`.  See issue #172 for related discussion, though it's about 307 temporary redirects.

The destination endpoint is not visible anywhere in the exception message. One needs to access the HTTP response body and then parse it. Here's a rudimentary example for illustration:

```
begin
  obj.put('Stuff')
rescue AWS::S3::Errors::PermanentRedirect => e
  new_endpoint = e.http_response.body.match(/<Endpoint>(.*?)<\/Endpoint>/)[1]
end
```

The response is XML, so an XML parser would also work. The endpoint hostnames I've received are all bucket-specific, not region-wide, like `the.bucket.name.s3.amazonaws.com` or `the.bucket.name.s3-external-3.amazonaws.com`.

To get past this case, re-instantiate your `AWS::S3` object with an `s3_endpoint` argument (applied [here](https://github.com/aws/aws-sdk-ruby/blob/master/lib/aws/s3/s3_object.rb#L1482)) and then retry the upload.

Finally, [This post](http://ruby.awsblog.com/post/TxVOTODBPHAEP9/Working-with-Regions) notes that `<svc>_endpoint` (like `s3_endpoint`) is deprecated in favor of providing `region`, but I don't see a region anywhere in the HTTP response, only an endpoint. Further, the example snippet in that post uses an `endpoint` argument as part of the newly-supported behavior. As of `aws-sdk` 1.37.0, passing the `s3_endpoint` config option leads to `s3.config.s3_endpoint` being defined, but passing `endpoint` still seems to make it use the default endpoint `s3.amazonaws.com`. I haven't explored this.

Yay! Uploads work! Well, most of the time. Now to part 2.

## Handling PermanentRedirects, part 2

The above change doesn't solve all uploads to all buckets. With our sample of buckets, it solves a large majority, but not nearly all. Because the buckets are user-supplied, we don't know a lot about how they were created and don't have a pattern about which ones don't work (or even whether the problem is transient for a given bucket). As I write this, this problem is still fresh for us and we're still troubleshooting it.

So, what happens? Using an example bucket of `bucket.example.com`, the initial request to the default endpoint redirects to `bucket.example.com.s3.amazonaws.com`. The second request, made with an S3 instance instantiated with `s3_endpoint` and where `customer_s3.config.s3_endpoint == 'bucket.example.com.s3.amazonaws.com'` and debug and `http_wire_trace` enabled, encounters a mismatch between the hostname and the presented SSL cert:

```
opening connection to bucket.example.com.s3.amazonaws.com...
opened
Conn close because of connect error hostname does not match the server certificate
opening connection to bucket.example.com.s3.amazonaws.com...
opened
Conn close because of connect error hostname does not match the server certificate
opening connection to bucket.example.com.s3.amazonaws.com...
opened
Conn close because of connect error hostname does not match the server certificate
opening connection to bucket.example.com.s3.amazonaws.com...
opened
Conn close because of connect error hostname does not match the server certificate
```

The configured number of retries are generated, then this is logged:

```
aws_sdk: put_object: service=S3 region=example.com.s3 
  host=bucket.example.com.s3.amazonaws.com
  uri=/bucket.example.com/abc/def/test_file.txt status=200 duration=4.187538
  retries=3 error=OpenSSL::SSL::SSLError 
  error_message=hostname does not match the server certificate
```

I can see that the `s3.config.ssl_ca_file` bundled with the gem exists and is readable. I haven't seen any other hostname mismatches. There's no transparent firewall or proxy in front of the systems making the request, and they successfully upload thousands of files every day. I can hit https://bucket.example.com.s3.amazonaws.com in a browser and see that it's presenting a wildcard cert. Basically, I have no explanation why this request doesn't succeed, other than that at least with the visibility I have, it seems to only happen some of the time that an alternate endpoint is used.

Just for grins, I tried disabling peer verification and I get a `AWS::S3::Errors::SignatureDoesNotMatch` ("The request signature we calculated does not match the signature you provided. Check your key and signing method.") that I also can't explain. Other requests succeed just fine, like enumerating the buckets with `s3.buckets`. Other uploads also succeed just fine, which makes me suspect it's a byproduct of something to do with the hostname or endpoint in the string being signed; that is, a symptom rather than anything to do with the actual problem.

## Status

That's where I am now. Handling PermanentRedirects in the caller by re-instantiating the `S3` instance with an `s3_endpoint` solves some cases, but not others, and I don't know what allows uploads to complete for the others.

If you have any suggestions, @trevorrowe, I'd be happy to document the rest of how to do this.

**Comment by trevorrowe** [comment]
@troy I've left this issue open for quite a while now. You are correct about the limitations in the SDK for dealing with dotted bucket names and the classic S3 endpoint (s3.amazonaws.com). The SDK is unable to use DNS to resolve to the proper region because it cannot verify the peer certificate for buckets that contain dots. This makes the Amazon S3 error unhelpful, as it is sending us back to the domain name we are unable to verify. The only solutions I am aware of currently are:

* Disable SSL (**not recommended**)
* Make a get bucket location request to determine the proper region to send the request to
* Make a request against the `:s3_endpoint` of s3-external-1.amazonaws.com. This is one of the public endpoints for the us-east-1 region.  Unlike other regions, this will return a temporary redirect instead of a permanent. This allows you to extract the actual region from the location header.

I've used the 3rd strategy in the V2 Ruby SDK to help deal with the new sigv4 only region eu-central-1. In order to correct marshall requests to that endpoint, you must know the region name ahead of time. I use s3-external-1.amazonaws.com to determine the region as mentioned above.

It is not a pretty work-around, but its the only thing that I have found that works.

I'm going to close this issue for now. I currently have no plans to modify the behvior for the v1 SDK at this time, but will consider adding this to the v2 AWS SDK for Ruby (https://github.com/aws/aws-sdk-core-ruby).
